File: conduit/tests/python/test_fx_coverage.py
"""Roundtrip encode/decode tests for FX block generated Python modules.

Covers:
- bitmap_fx: bitmap-presence struct inside an FX-extended message
- empty_fx: FX block with no extension fields
- fx_ia5_string: FX block with IA5-encoded string and uint16
- fx_string: FX block with ASCII string, bytes, and uint16
"""
import pytest


# ============================================================================
# bitmap_fx: BitmapItems (FSPEC bitmap) inside Category message
# ============================================================================


class TestBitmapFxAllPresent:
    """All three bitmap items are present."""

    def test_roundtrip_all_items(self):
        from bitmap_fx import Category
        from bitmap_fx.structs import BitmapItems, DataItem010, DataItem020, DataItem030

        msg = Category()
        items = BitmapItems()
        item010 = DataItem010()
        item010.sac = 0xAA
        item010.sic = 0xBB
        items.item010 = item010

        item020 = DataItem020()
        item020.code = 0x1234
        items.item020 = item020

        item030 = DataItem030()
        item030.value = 0xDEADBEEF
        items.item030 = item030

        msg.items = items

        data = msg.encode_bytes()
        msg2 = Category.decode_bytes(data)
        assert msg2.items.item010.sac == 0xAA
        assert msg2.items.item010.sic == 0xBB
        assert msg2.items.item020.code == 0x1234
        assert msg2.items.item030.value == 0xDEADBEEF

    def test_double_encode_stability(self):
        from bitmap_fx import Category
        from bitmap_fx.structs import BitmapItems, DataItem010, DataItem020, DataItem030

        msg = Category()
        items = BitmapItems()
        item010 = DataItem010()
        item010.sac = 1
        item010.sic = 2
        items.item010 = item010
        item020 = DataItem020()
        item020.code = 300
        items.item020 = item020
        item030 = DataItem030()
        item030.value = 400
        items.item030 = item030
        msg.items = items

        data1 = msg.encode_bytes()
        msg2 = Category.decode_bytes(data1)
        data2 = msg2.encode_bytes()
        assert data1 == data2


class TestBitmapFxPartialPresence:
    """Only some bitmap items are present."""

    def test_only_item010(self):
        from bitmap_fx import Category
        from bitmap_fx.structs import BitmapItems, DataItem010

        msg = Category()
        items = BitmapItems()
        item010 = DataItem010()
        item010.sac = 0x10
        item010.sic = 0x20
        items.item010 = item010
        msg.items = items

        data = msg.encode_bytes()
        msg2 = Category.decode_bytes(data)
        assert msg2.items.item010.sac == 0x10
        assert msg2.items.item010.sic == 0x20
        assert msg2.items.item020 is None
        assert msg2.items.item030 is None

    def test_only_item030(self):
        from bitmap_fx import Category
        from bitmap_fx.structs import BitmapItems, DataItem030

        msg = Category()
        items = BitmapItems()
        item030 = DataItem030()
        item030.value = 0xCAFEBABE
        items.item030 = item030
        msg.items = items

        data = msg.encode_bytes()
        msg2 = Category.decode_bytes(data)
        assert msg2.items.item010 is None
        assert msg2.items.item020 is None
        assert msg2.items.item030.value == 0xCAFEBABE


class TestBitmapFxNoItems:
    """No bitmap items present."""

    def test_empty_bitmap(self):
        from bitmap_fx import Category
        from bitmap_fx.structs import BitmapItems

        msg = Category()
        items = BitmapItems()
        msg.items = items

        data = msg.encode_bytes()
        msg2 = Category.decode_bytes(data)
        assert msg2.items.item010 is None
        assert msg2.items.item020 is None
        assert msg2.items.item030 is None


class TestBitmapFxBoundaryValues:
    """Boundary values for bitmap field types."""

    def test_max_values(self):
        from bitmap_fx import Category
        from bitmap_fx.structs import BitmapItems, DataItem010, DataItem020, DataItem030

        msg = Category()
        items = BitmapItems()
        item010 = DataItem010()
        item010.sac = 0xFF
        item010.sic = 0xFF
        items.item010 = item010
        item020 = DataItem020()
        item020.code = 0xFFFF
        items.item020 = item020
        item030 = DataItem030()
        item030.value = 0xFFFFFFFF
        items.item030 = item030
        msg.items = items

        data = msg.encode_bytes()
        msg2 = Category.decode_bytes(data)
        assert msg2.items.item010.sac == 0xFF
        assert msg2.items.item010.sic == 0xFF
        assert msg2.items.item020.code == 0xFFFF
        assert msg2.items.item030.value == 0xFFFFFFFF

    def test_zero_values(self):
        from bitmap_fx import Category
        from bitmap_fx.structs import BitmapItems, DataItem010, DataItem020, DataItem030

        msg = Category()
        items = BitmapItems()
        item010 = DataItem010()
        item010.sac = 0
        item010.sic = 0
        items.item010 = item010
        item020 = DataItem020()
        item020.code = 0
        items.item020 = item020
        item030 = DataItem030()
        item030.value = 0
        items.item030 = item030
        msg.items = items

        data = msg.encode_bytes()
        msg2 = Category.decode_bytes(data)
        assert msg2.items.item010.sac == 0
        assert msg2.items.item010.sic == 0
        assert msg2.items.item020.code == 0
        assert msg2.items.item030.value == 0


# ============================================================================
# empty_fx: EmptyFxMsg has header + FX with no extension fields
# ============================================================================


class TestEmptyFxRoundtrip:
    """EmptyFxMsg has a header byte and an empty FX block."""

    def test_header_roundtrip(self):
        from empty_fx import EmptyFxMsg

        msg = EmptyFxMsg()
        msg.header = 0x42

        data = msg.encode_bytes()
        msg2 = EmptyFxMsg.decode_bytes(data)
        assert msg2.header == 0x42

    def test_header_zero(self):
        from empty_fx import EmptyFxMsg

        msg = EmptyFxMsg()
        msg.header = 0

        data = msg.encode_bytes()
        msg2 = EmptyFxMsg.decode_bytes(data)
        assert msg2.header == 0

    def test_header_max(self):
        from empty_fx import EmptyFxMsg

        msg = EmptyFxMsg()
        msg.header = 0xFF

        data = msg.encode_bytes()
        msg2 = EmptyFxMsg.decode_bytes(data)
        assert msg2.header == 0xFF

    def test_double_encode_stability(self):
        from empty_fx import EmptyFxMsg

        msg = EmptyFxMsg()
        msg.header = 0xAB

        data1 = msg.encode_bytes()
        msg2 = EmptyFxMsg.decode_bytes(data1)
        data2 = msg2.encode_bytes()
        assert data1 == data2


# ============================================================================
# fx_ia5_string: FxIa5Msg with header + FX(label:ia5 string 8, code:uint16)
# ============================================================================


class TestFxIa5StringWithExtension:
    """FxIa5Msg with FX items present."""

    def test_roundtrip_with_fx(self):
        from fx_ia5_string import FxIa5Msg

        msg = FxIa5Msg()
        msg.header = 0x10
        msg.label = "HELLO"
        msg.code = 0xABCD

        data = msg.encode_bytes()
        msg2 = FxIa5Msg.decode_bytes(data)
        assert msg2.header == 0x10
        assert msg2.label.rstrip() == "HELLO"
        assert msg2.code == 0xABCD

    def test_full_length_label(self):
        from fx_ia5_string import FxIa5Msg

        msg = FxIa5Msg()
        msg.header = 0x20
        msg.label = "ABCDEFGH"  # exactly 8 chars
        msg.code = 1000

        data = msg.encode_bytes()
        msg2 = FxIa5Msg.decode_bytes(data)
        assert msg2.label == "ABCDEFGH"
        assert msg2.code == 1000

    def test_empty_label(self):
        from fx_ia5_string import FxIa5Msg

        msg = FxIa5Msg()
        msg.header = 0x30
        msg.label = ""
        msg.code = 0

        data = msg.encode_bytes()
        msg2 = FxIa5Msg.decode_bytes(data)
        assert msg2.label.rstrip() == ""
        assert msg2.code == 0

    def test_double_encode_stability(self):
        from fx_ia5_string import FxIa5Msg

        msg = FxIa5Msg()
        msg.header = 0x77
        msg.label = "TEST"
        msg.code = 9999

        data1 = msg.encode_bytes()
        msg2 = FxIa5Msg.decode_bytes(data1)
        data2 = msg2.encode_bytes()
        assert data1 == data2


class TestFxIa5StringNoExtension:
    """FxIa5Msg without FX items (header only)."""

    def test_no_fx_items(self):
        from fx_ia5_string import FxIa5Msg

        msg = FxIa5Msg()
        msg.header = 0xFF

        data = msg.encode_bytes()
        msg2 = FxIa5Msg.decode_bytes(data)
        assert msg2.header == 0xFF
        assert msg2.label is None
        assert msg2.code is None


class TestFxIa5StringBoundary:
    """Boundary values for FxIa5Msg fields."""

    def test_code_max_value(self):
        from fx_ia5_string import FxIa5Msg

        msg = FxIa5Msg()
        msg.header = 0x01
        msg.label = "MAX"
        msg.code = 0xFFFF

        data = msg.encode_bytes()
        msg2 = FxIa5Msg.decode_bytes(data)
        assert msg2.code == 0xFFFF


# ============================================================================
# fx_string: FxStringMsg with header + FX(label:string 10, payload:bytes 4,
#            extra:uint16)
# ============================================================================


class TestFxStringWithExtension:
    """FxStringMsg with FX items present."""

    def test_roundtrip_with_fx(self):
        from fx_string import FxStringMsg

        msg = FxStringMsg()
        msg.header = 0x55
        msg.label = "TestLabel"
        msg.payload = b'\xDE\xAD\xBE\xEF'
        msg.extra = 0x1234

        data = msg.encode_bytes()
        msg2 = FxStringMsg.decode_bytes(data)
        assert msg2.header == 0x55
        assert msg2.label.rstrip('\x00') == "TestLabel"
        assert msg2.payload == b'\xDE\xAD\xBE\xEF'
        assert msg2.extra == 0x1234

    def test_full_length_label(self):
        from fx_string import FxStringMsg

        msg = FxStringMsg()
        msg.header = 0xAA
        msg.label = "0123456789"  # exactly 10 chars
        msg.payload = b'\x01\x02\x03\x04'
        msg.extra = 42

        data = msg.encode_bytes()
        msg2 = FxStringMsg.decode_bytes(data)
        assert msg2.label == "0123456789"

    def test_empty_strings_and_zero_payload(self):
        from fx_string import FxStringMsg

        msg = FxStringMsg()
        msg.header = 0x00
        msg.label = ""
        msg.payload = b'\x00\x00\x00\x00'
        msg.extra = 0

        data = msg.encode_bytes()
        msg2 = FxStringMsg.decode_bytes(data)
        assert msg2.label.rstrip('\x00') == ""
        assert msg2.payload == b'\x00\x00\x00\x00'
        assert msg2.extra == 0

    def test_double_encode_stability(self):
        from fx_string import FxStringMsg

        msg = FxStringMsg()
        msg.header = 0x33
        msg.label = "Stable"
        msg.payload = b'\xCA\xFE\xBA\xBE'
        msg.extra = 0x7FFF

        data1 = msg.encode_bytes()
        msg2 = FxStringMsg.decode_bytes(data1)
        data2 = msg2.encode_bytes()
        assert data1 == data2


class TestFxStringNoExtension:
    """FxStringMsg without FX items."""

    def test_no_fx_items(self):
        from fx_string import FxStringMsg

        msg = FxStringMsg()
        msg.header = 0xBB

        data = msg.encode_bytes()
        msg2 = FxStringMsg.decode_bytes(data)
        assert msg2.header == 0xBB
        assert msg2.label is None
        assert msg2.payload is None
        assert msg2.extra is None


class TestFxStringBoundary:
    """Boundary values for FxStringMsg fields."""

    def test_extra_max_value(self):
        from fx_string import FxStringMsg

        msg = FxStringMsg()
        msg.header = 0x01
        msg.label = "X"
        msg.payload = b'\xFF\xFF\xFF\xFF'
        msg.extra = 0xFFFF

        data = msg.encode_bytes()
        msg2 = FxStringMsg.decode_bytes(data)
        assert msg2.extra == 0xFFFF
        assert msg2.payload == b'\xFF\xFF\xFF\xFF'

File: conduit/tests/java/TestFxCoverage.java
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import static org.junit.jupiter.api.Assertions.*;

import fx_block.FxBlockMsg;

import fx_ia5_string.FxIa5Msg;
import fx_string.FxStringMsg;

public class TestFxCoverage {

    // ========== FxBlockMsg Tests ==========

    @Test
    @DisplayName("FxBlockMsg: basic roundtrip with all fields set")
    void testFxBlockMsgBasicRoundtrip() {
        FxBlockMsg msg = new FxBlockMsg();
        msg.header = 0x01;
        msg.fxField1 = (short) 100;
        msg.fxField2 = (short) 200;

        byte[] data = msg.encodeBytes();
        FxBlockMsg decoded = FxBlockMsg.decodeBytes(data);

        assertEquals(msg.header, decoded.header);
        assertEquals(msg.fxField1, decoded.fxField1);
        assertEquals(msg.fxField2, decoded.fxField2);
    }

    @Test
    @DisplayName("FxBlockMsg: optional FX fields can be null")
    void testFxBlockMsgOptionalFieldsNull() {
        FxBlockMsg msg = new FxBlockMsg();
        msg.header = 0x02;
        msg.fxField1 = null;
        msg.fxField2 = null;

        byte[] data = msg.encodeBytes();
        FxBlockMsg decoded = FxBlockMsg.decodeBytes(data);

        assertEquals(msg.header, decoded.header);
        assertNull(decoded.fxField1);
        assertNull(decoded.fxField2);
    }

    @Test
    @DisplayName("FxBlockMsg: partial FX fields - first set, second null")
    void testFxBlockMsgPartialFields() {
        FxBlockMsg msg = new FxBlockMsg();
        msg.header = 0x03;
        msg.fxField1 = (short) 42;
        msg.fxField2 = null;

        byte[] data = msg.encodeBytes();
        FxBlockMsg decoded = FxBlockMsg.decodeBytes(data);

        assertEquals(msg.header, decoded.header);
        assertEquals((short) 42, decoded.fxField1);
        assertNull(decoded.fxField2);
    }

    @Test
    @DisplayName("FxBlockMsg: boundary value 0xFF in FX field")
    void testFxBlockMsgBoundary0xFF() {
        FxBlockMsg msg = new FxBlockMsg();
        msg.header = 0x04;
        msg.fxField1 = (short) 0xFF;
        msg.fxField2 = null;

        byte[] data = msg.encodeBytes();
        FxBlockMsg decoded = FxBlockMsg.decodeBytes(data);

        assertEquals((short) 0xFF, decoded.fxField1);
    }

    @Test
    @DisplayName("FxBlockMsg: double-encode stability")
    void testFxBlockMsgDoubleEncode() {
        FxBlockMsg msg = new FxBlockMsg();
        msg.header = 0x05;
        msg.fxField1 = (short) 50;
        msg.fxField2 = (short) 60;

        byte[] data1 = msg.encodeBytes();
        FxBlockMsg decoded1 = FxBlockMsg.decodeBytes(data1);
        byte[] data2 = decoded1.encodeBytes();

        assertArrayEquals(data1, data2);
    }

    // ========== FxIa5Msg Tests ==========

    @Test
    @DisplayName("FxIa5Msg: roundtrip with IA5 string and uint16")
    void testFxIa5MsgRoundtrip() {
        FxIa5Msg msg = new FxIa5Msg();
        msg.header = 0x20;
        msg.ia5Value = "ABCDEFGH";
        msg.numValue = 1234;

        byte[] data = msg.encodeBytes();
        FxIa5Msg decoded = FxIa5Msg.decodeBytes(data);

        assertEquals(msg.header, decoded.header);
        assertEquals("ABCDEFGH", decoded.ia5Value.trim());
        assertEquals(1234, decoded.numValue);
    }

    @Test
    @DisplayName("FxIa5Msg: IA5 string padding and trimming")
    void testFxIa5MsgStringPadding() {
        FxIa5Msg msg = new FxIa5Msg();
        msg.header = 0x21;
        msg.ia5Value = "AB";
        msg.numValue = 0;

        byte[] data = msg.encodeBytes();
        FxIa5Msg decoded = FxIa5Msg.decodeBytes(data);

        assertEquals("AB", decoded.ia5Value.trim());
    }

    @Test
    @DisplayName("FxIa5Msg: uint16 boundary 0xFFFF")
    void testFxIa5MsgBoundary0xFFFF() {
        FxIa5Msg msg = new FxIa5Msg();
        msg.header = 0x22;
        msg.ia5Value = "TESTTEST";
        msg.numValue = 0xFFFF;

        byte[] data = msg.encodeBytes();
        FxIa5Msg decoded = FxIa5Msg.decodeBytes(data);

        assertEquals(0xFFFF, decoded.numValue);
    }

    @Test
    @DisplayName("FxIa5Msg: double-encode stability")
    void testFxIa5MsgDoubleEncode() {
        FxIa5Msg msg = new FxIa5Msg();
        msg.header = 0x23;
        msg.ia5Value = "ROUNDTRP";
        msg.numValue = 500;

        byte[] data1 = msg.encodeBytes();
        FxIa5Msg decoded1 = FxIa5Msg.decodeBytes(data1);
        byte[] data2 = decoded1.encodeBytes();

        assertArrayEquals(data1, data2);
    }

    // ========== FxStringMsg Tests ==========

    @Test
    @DisplayName("FxStringMsg: roundtrip with ASCII string, bytes, and uint16")
    void testFxStringMsgRoundtrip() {
        FxStringMsg msg = new FxStringMsg();
        msg.header = 0x30;
        msg.strValue = "ABCDEFGHIJ";
        msg.bytesValue = new byte[]{0x01, 0x02, 0x03, 0x04};
        msg.numValue = 9999;

        byte[] data = msg.encodeBytes();
        FxStringMsg decoded = FxStringMsg.decodeBytes(data);

        assertEquals(msg.header, decoded.header);
        assertEquals("ABCDEFGHIJ", decoded.strValue.trim());
        assertArrayEquals(new byte[]{0x01, 0x02, 0x03, 0x04}, decoded.bytesValue);
        assertEquals(9999, decoded.numValue);
    }

    @Test
    @DisplayName("FxStringMsg: short string is padded correctly")
    void testFxStringMsgShortStringPadding() {
        FxStringMsg msg = new FxStringMsg();
        msg.header = 0x31;
        msg.strValue = "HI";
        msg.bytesValue = new byte[]{0x00, 0x00, 0x00, 0x00};
        msg.numValue = 0;

        byte[] data = msg.encodeBytes();
        FxStringMsg decoded = FxStringMsg.decodeBytes(data);

        assertEquals("HI", decoded.strValue.trim());
    }

    @Test
    @DisplayName("FxStringMsg: boundary values for bytes and uint16")
    void testFxStringMsgBoundaryValues() {
        FxStringMsg msg = new FxStringMsg();
        msg.header = 0x32;
        msg.strValue = "BOUNDARY01";
        msg.bytesValue = new byte[]{(byte) 0xFF, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF};
        msg.numValue = 0xFFFF;

        byte[] data = msg.encodeBytes();
        FxStringMsg decoded = FxStringMsg.decodeBytes(data);

        assertArrayEquals(new byte[]{(byte) 0xFF, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF}, decoded.bytesValue);
        assertEquals(0xFFFF, decoded.numValue);
    }

    @Test
    @DisplayName("FxStringMsg: zero uint16 value")
    void testFxStringMsgZeroNum() {
        FxStringMsg msg = new FxStringMsg();
        msg.header = 0x33;
        msg.strValue = "ZEROTEST00";
        msg.bytesValue = new byte[]{0x0A, 0x0B, 0x0C, 0x0D};
        msg.numValue = 0;

        byte[] data = msg.encodeBytes();
        FxStringMsg decoded = FxStringMsg.decodeBytes(data);

        assertEquals(0, decoded.numValue);
    }

    @Test
    @DisplayName("FxStringMsg: double-encode stability")
    void testFxStringMsgDoubleEncode() {
        FxStringMsg msg = new FxStringMsg();
        msg.header = 0x34;
        msg.strValue = "STABLE0001";
        msg.bytesValue = new byte[]{0x11, 0x22, 0x33, 0x44};
        msg.numValue = 777;

        byte[] data1 = msg.encodeBytes();
        FxStringMsg decoded1 = FxStringMsg.decodeBytes(data1);
        byte[] data2 = decoded1.encodeBytes();

        assertArrayEquals(data1, data2);
    }
}
